source/lac/petsc_communication_pattern.cc

```diff
@@ -0,0 +1,516 @@
+// ---------------------------------------------------------------------
+//
+// Copyright (C) 2023 by the deal.II authors
+//
+// This file is part of the deal.II library.
+//
+// The deal.II library is free software; you can use it, redistribute
+// it, and/or modify it under the terms of the GNU Lesser General
+// Public License as published by the Free Software Foundation; either
+// version 2.1 of the License, or (at your option) any later version.
+// The full text of the license can be found in the file LICENSE.md at
+// the top level directory of deal.II.
+//
+// ---------------------------------------------------------------------
+
+#include <deal.II/lac/petsc_communication_pattern.h>
+
+#ifdef DEAL_II_WITH_PETSC
+
+#  include <deal.II/base/mpi.h>
+
+#  include <deal.II/lac/exceptions.h>
+
+DEAL_II_NAMESPACE_OPEN
+// Shorthand notation for PETSc error codes.
+#  define AssertPETSc(code)                              \
+    do                                                   \
+      {                                                  \
+        PetscErrorCode __ierr = (code);                  \
+        AssertThrow(__ierr == 0, ExcPETScError(__ierr)); \
+      }                                                  \
+    while (0)
+
+namespace PETScWrappers
+{
+  CommunicationPattern::CommunicationPattern()
+    : sf(nullptr)
+  {}
+
+  CommunicationPattern::~CommunicationPattern()
+  {
+    clear();
+  }
+
+  void
+  CommunicationPattern::reinit(const types::global_dof_index local_size,
+                               const IndexSet &              ghost_indices,
+                               const MPI_Comm &              communicator)
+  {
+    clear();
+
+    PetscLayout layout;
+    AssertPETSc(PetscLayoutCreate(communicator, &layout));
+    AssertPETSc(PetscLayoutSetLocalSize(layout, local_size));
+    AssertPETSc(PetscLayoutSetUp(layout));
+
+    PetscInt start, end;
+    AssertPETSc(PetscLayoutGetRange(layout, &start, &end));
+
+    IndexSet want;
+    want.add_range(start, end);
+    want.add_indices(ghost_indices);
+    want.compress();
+
+    const PetscInt *idxs;
+    PetscInt        n;
+    IS              is = want.make_petsc_is(communicator);
+    AssertPETSc(ISGetLocalSize(is, &n));
+    AssertPETSc(ISGetIndices(is, &idxs));
+
+    AssertPETSc(PetscSFCreate(communicator, &sf));
+    AssertPETSc(
+      PetscSFSetGraphLayout(sf, layout, n, nullptr, PETSC_OWN_POINTER, idxs));
+    AssertPETSc(PetscSFSetUp(sf));
+
+    AssertPETSc(ISRestoreIndices(is, &idxs));
+    AssertPETSc(ISDestroy(&is));
+    AssertPETSc(PetscLayoutDestroy(&layout));
+  }
+
+  void
+  CommunicationPattern::reinit(const IndexSet &vector_space_vector_index_set,
+                               const IndexSet &read_write_vector_index_set,
```

Comment:
These argument names are not particularly good (but not your fault). If you were so inclined, would you use the names of #14740 ?

Reply:
Will do